core/include/traccc/utils/type_traits.hpp

```diff
@@ -0,0 +1,32 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2021-2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+namespace traccc::details {
+
+/// Helper trait for detecting when a type is a non-const version of another
+///
+/// This comes into play multiple times to enable certain constructors
+/// conditionally through SFINAE.
+///
+template <typename CTYPE, typename NCTYPE>
+struct is_same_nc {
+    static constexpr bool value = false;
+};
+
+template <typename TYPE>
+struct is_same_nc<TYPE, TYPE> {
+    static constexpr bool value = true;
+};
+
+template <typename TYPE>
+struct is_same_nc<const TYPE, TYPE> {
+    static constexpr bool value = true;
+};
```

Comment:
Even if we keep it this, there are a few problems with it that we might want to consider:

1. This does not properly handle the case where the first type is non-const and the second type is const.
2. This breaks down if we are talking about references or pointers, in which case the reference or pointer type constructor is applied after the const type constructor, and must be stripped from it. Although that may be intended.

Reply:
This was the intention... The trait should "only work in one direction".

Reply:
Then I would argue it is very confusingly named. :fearful: